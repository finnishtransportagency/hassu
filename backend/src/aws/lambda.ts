import { Lambda } from "@aws-sdk/client-lambda";
import { log } from "../logger";

const lambda: Lambda = new Lambda({});

export async function invokeLambda(functionName: string, synchronousCall: boolean, payload?: string): Promise<string | undefined> {
  try {
    if (synchronousCall) {
      const output = await lambda.invoke({
        FunctionName: functionName,
        Payload: new TextEncoder().encode(payload || "{}"),
      });
      const outputPayload = output.Payload;
      return new TextDecoder("utf-8").decode(outputPayload);
    } else {
      await lambda.invokeAsync({
        FunctionName: functionName,
        InvokeArgs: payload || "{}",
      });
    }
  } catch (e) {
    log.error("invokeLambda", e);
    throw e;
  }
}
