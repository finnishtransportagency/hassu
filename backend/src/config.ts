import { log } from "./logger";
import * as AWSXRay from "aws-xray-sdk";

const config = {
  projektiTableName: process.env.TABLE_PROJEKTI,
  projektiArchiveTableName: process.env.TABLE_PROJEKTI_ARCHIVE,
  cognitoURL: process.env.COGNITO_URL,
  velhoAuthURL: process.env.VELHO_AUTH_URL,
  velhoApiURL: process.env.VELHO_API_URL,
  velhoUsername: process.env.VELHO_USERNAME,
  velhoPassword: process.env.VELHO_PASSWORD,

  personSearchApiURL: process.env.PERSON_SEARCH_API_URL,
  personSearchApiURLProd: process.env.PERSON_SEARCH_API_URL_PROD,
  personSearchUsername: process.env.PERSON_SEARCH_API_USERNAME,
  personSearchPassword: process.env.PERSON_SEARCH_API_PASSWORD,
  personSearchAccountTypes: process.env.PERSON_SEARCH_API_ACCOUNT_TYPES?.split(","),
  personSearchUpdaterLambdaArn: process.env.PERSON_SEARCH_UPDATER_LAMBDA_ARN || "",

  searchDomain: process.env.SEARCH_DOMAIN,

  env: process.env.ENVIRONMENT,

  frontendDomainName: process.env.FRONTEND_DOMAIN_NAME,

  frontendPrivateKey: process.env.FRONTEND_PRIVATEKEY,

  uploadBucketName: process.env.UPLOAD_BUCKET_NAME,
  yllapitoBucketName: process.env.YLLAPITO_BUCKET_NAME,
  publicBucketName: process.env.PUBLIC_BUCKET_NAME,
  internalBucketName: process.env.INTERNAL_BUCKET_NAME,
  archiveBucketName: process.env.ARCHIVE_BUCKET_NAME,

  smtpKeyId: process.env.SMTP_KEY_ID,
  smtpSecret: process.env.SMTP_SECRET,

  emailsOn: process.env.EMAILS_ON,
  emailsTo: process.env.EMAILS_TO,
  isProd: (): boolean => process.env.ENVIRONMENT == "prod",
};

process.env.AWS_XRAY_CONTEXT_MISSING = "LOG_ERROR";
AWSXRay.setContextMissingStrategy("IGNORE_ERROR");

// Create your own logger, or instantiate one using a library.
AWSXRay.setLogger({
  error: (message, meta) => {
    log.error(message, { meta });
  },
  warn: (message, meta) => {
    log.warn(message, { meta });
  },
  info: (message, meta) => {
    // Reduce unnecessary logging in local development
    if (process.env.NODE_ENV !== "development") {
      log.info(message, { meta });
    }
  },
  debug: () => {
    // Ignore
  },
});

export { config };
