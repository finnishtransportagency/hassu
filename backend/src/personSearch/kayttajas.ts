import { Kayttaja } from "../../../common/graphql/apiModel";

export class Kayttajas {
  kayttajaMap: Record<string, Kayttaja>;

  constructor(kayttajaMap: Record<string, Kayttaja>) {
    this.kayttajaMap = kayttajaMap;
  }

  getKayttajaByUid(uid: string | undefined): Kayttaja | undefined {
    if (uid) {
      return this.kayttajaMap[uid];
    }
  }

  findByEmail(email: string): Kayttaja | undefined {
    for (const kayttaja of Object.values(this.kayttajaMap)) {
      if (kayttaja.email === email) {
        return kayttaja;
      }
    }
  }

  static fromKayttajaList(kayttajas: Kayttaja[]) {
    return new Kayttajas(
      kayttajas.reduce((map, kayttaja) => {
        if (kayttaja.uid) {
          map[kayttaja.uid] = kayttaja;
        }
        return map;
      }, {} as Record<string, Kayttaja>)
    );
  }

  asList() {
    const list = [];
    for (const uid in this.kayttajaMap) {
      if (this.kayttajaMap.hasOwnProperty(uid)) {
        list.push(this.kayttajaMap[uid]);
      }
    }
    return list;
  }

  asMap() {
    return this.kayttajaMap;
  }

  findByText(hakusana: string): Kayttaja[] {
    if (hakusana.length >= 3) {
      return Object.values(this.kayttajaMap).reduce((list, kayttaja) => {
        if (kayttaja.etuNimi.includes(hakusana) || kayttaja.sukuNimi.includes(hakusana)) {
          list.push(kayttaja);
        }
        return list;
      }, [] as Kayttaja[]);
    }
    return [];
  }
}
