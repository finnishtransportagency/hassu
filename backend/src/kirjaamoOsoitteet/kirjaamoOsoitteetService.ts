import { GetParameterResult } from "aws-sdk/clients/ssm";
import log from "loglevel";
import { KirjaamoOsoite } from "../../../common/graphql/apiModel";
import { getSSM } from "../aws/client";

async function listKirjaamoOsoitteet(): Promise<KirjaamoOsoite[]> {
  const parameterName = "/kirjaamoOsoitteet";
  let kirjaamoOsoitteet: KirjaamoOsoite[] = [];
  try {
    const response: GetParameterResult = await getSSM().getParameter({ Name: parameterName }).promise();
    kirjaamoOsoitteet = response.Parameter?.Value ? JSON.parse(response.Parameter.Value) : [];
  } catch (e) {
    log.error(`Could not read or parse 'kirjaamoOsoitteet' from SSM`, e);
  }
  return kirjaamoOsoitteet;
}

export const kirjaamoOsoitteetService = { listKirjaamoOsoitteet };
