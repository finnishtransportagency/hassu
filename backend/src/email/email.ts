import * as nodemailer from "nodemailer";
import { config } from "../config";
import { log } from "../logger";
import { MailOptions } from "nodemailer/lib/smtp-transport";
import cloneDeep from "lodash/cloneDeep";

const transporter = nodemailer.createTransport({
  port: 465,
  host: "email-smtp.eu-west-1.amazonaws.com",
  secure: true,
  auth: {
    user: config.smtpKeyId,
    pass: config.smtpSecret,
  },
  debug: true,
});

export type EmailOptions = Pick<MailOptions, "to" | "subject" | "text" | "attachments">;

async function sendEmail(options: EmailOptions): Promise<void> {
  if (config.emailsOn !== "true") {
    const { attachments: attachments, ...restOptions } = options;
    log.info("Sähköpostin lähetys kytketty pois päältä", {
      emailOptions: restOptions,
      attachments: attachments?.map((att) => {
        const a = cloneDeep(att);
        a.content = "***test***";
        return a;
      }),
    });
    return;
  }
  try {
    const messageInfo = await transporter.sendMail({
      from: "noreply.hassu@vaylapilvi.fi",
      ...options,
      to: config.emailsTo || options.to,
    });
    log.info("Email lähetetty", messageInfo);
  } catch (e) {
    log.error("Email lähetys epäonnistui", e);
  }
}

export const emailClient = { sendEmail };
