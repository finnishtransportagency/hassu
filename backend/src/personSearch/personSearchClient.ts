import { Kayttaja, ProjektiRooli } from "../../../common/graphql/apiModel";
import { DBVaylaUser } from "../database/model/projekti";
import { isAorL } from "../user";
import { mergeKayttaja } from "./personAdapter";
import { personSearchUpdaterClient } from "./personSearchUpdaterClient";
import { s3Cache } from "../cache/s3Cache";
import log from "loglevel";
import { Kayttajas } from "./kayttajas";
import { wrapXrayAsync } from "../aws/xray";

export const S3CACHE_TTL_MILLIS = 15 * 60 * 1000; // 15 min
export enum SearchMode {
  EMAIL,
  UID,
}

export const PERSON_SEARCH_CACHE_KEY = "users.json";

export class PersonSearchClient {
  public async getKayttajas(): Promise<Kayttajas> {
    try {
      return await wrapXrayAsync("getKayttajas", async () => {
        const kayttajaMap: Record<string, Kayttaja> = await s3Cache.get(
          PERSON_SEARCH_CACHE_KEY,
          S3CACHE_TTL_MILLIS,
          async () => {
            personSearchUpdaterClient.triggerUpdate();
          },
          async () => {
            return await personSearchUpdaterClient.readUsersFromSearchUpdaterLambda();
          }
        );
        return new Kayttajas(kayttajaMap);
      });
    } catch (e) {
      log.error("getKayttajas", e);
      throw e;
    }
  }

  public async fillInUserInfoFromUserManagement({
    user: user,
    searchMode,
  }: {
    user: Partial<DBVaylaUser>;
    searchMode: SearchMode;
  }): Promise<DBVaylaUser | undefined> {
    const kayttajas = await this.getKayttajas();
    let account: Kayttaja | undefined;

    if (searchMode === SearchMode.UID) {
      account = kayttajas.getKayttajaByUid(user.kayttajatunnus);
    } else {
      account = kayttajas.findByEmail(user.email);
    }
    if (account) {
      // Projektipaallikko must be either L or A account
      if (user.rooli === ProjektiRooli.PROJEKTIPAALLIKKO && !isAorL(account)) {
        return;
      }
      mergeKayttaja(user, account);
      return user as DBVaylaUser;
    }
    return;
  }
}

export const personSearch = new PersonSearchClient();
