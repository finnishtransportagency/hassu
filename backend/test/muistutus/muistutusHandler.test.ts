import { describe } from "mocha";
import * as sinon from "sinon";
import { ProjektiFixture } from "../fixture/projektiFixture";
import { UserFixture } from "../fixture/userFixture";
import { personSearch } from "../../src/personSearch/personSearchClient";
import { userService } from "../../src/user";
import { MuistutusInput } from "hassu-common/graphql/apiModel";
import { PersonSearchFixture } from "../personSearch/lambda/personSearchFixture";
import { Kayttajas } from "../../src/personSearch/kayttajas";
import { emailClient } from "../../src/email/email";
import { DBMuistuttaja, muistutusHandler } from "../../src/muistutus/muistutusHandler";
import { projektiDatabase } from "../../src/database/projektiDatabase";
import { kirjaamoOsoitteetService } from "../../src/kirjaamoOsoitteet/kirjaamoOsoitteetService";
import { S3Mock } from "../aws/awsMock";

import { assert, expect } from "chai";
import { mockClient } from "aws-sdk-client-mock";
import { BatchGetCommand, DynamoDBDocumentClient, GetCommand, PutCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { config } from "../../src/config";
import { identifyMockUser } from "../../src/user/userService";

describe("muistutusHandler", () => {
  const userFixture = new UserFixture(userService);
  new S3Mock();

  afterEach(() => {
    sinon.restore();
    userFixture.logout();
  });

  describe("handleEvent", () => {
    let fixture: ProjektiFixture;
    let personSearchFixture: PersonSearchFixture;
    let loadProjektiByOidStub: sinon.SinonStub;
    let sendEmailStub: sinon.SinonStub;
    let sendTurvapostiEmailStub: sinon.SinonStub;
    let getKayttajasStub: sinon.SinonStub;
    let kirjaamoOsoitteetStub: sinon.SinonStub;

    beforeEach(() => {
      getKayttajasStub = sinon.stub(personSearch, "getKayttajas");
      loadProjektiByOidStub = sinon.stub(projektiDatabase, "loadProjektiByOid");
      sendEmailStub = sinon.stub(emailClient, "sendEmail");
      sendTurvapostiEmailStub = sinon.stub(emailClient, "sendTurvapostiEmail");
      kirjaamoOsoitteetStub = sinon.stub(kirjaamoOsoitteetService, "listKirjaamoOsoitteet");

      fixture = new ProjektiFixture();
      personSearchFixture = new PersonSearchFixture();
      getKayttajasStub.resolves(
        Kayttajas.fromKayttajaList([
          personSearchFixture.pekkaProjari,
          personSearchFixture.mattiMeikalainen,
          personSearchFixture.manuMuokkaaja,
          personSearchFixture.createKayttaja("A2"),
        ])
      );
      sendEmailStub.resolves();
      loadProjektiByOidStub.resolves(fixture.dbProjekti3);
      kirjaamoOsoitteetStub.resolves([
        { nimi: "ETELA_POHJANMAAN_ELY", sahkoposti: "kirjaamo.etela-pohjanmaa@ely-keskus.fi" },
        { nimi: "ETELA_SAVO_ELY", sahkoposti: "kirjaamo.etela-savo@ely-keskus.fi" },
        { nimi: "HAME_ELY", sahkoposti: "kirjaamo.hame@ely-keskus.fi" },
        { nimi: "KAAKKOIS_SUOMEN_ELY", sahkoposti: "kirjaamo.kaakkois-suomi@ely-keskus.fi" },
        { nimi: "KAINUUN_ELY", sahkoposti: "kirjaamo.kainuu@ely-keskus.fi" },
        { nimi: "KESKI_SUOMEN_ELY", sahkoposti: "kirjaamo.keski-suomi@ely-keskus.fi" },
        { nimi: "LAPIN_ELY", sahkoposti: "kirjaamo.lappi@ely-keskus.fi" },
        { nimi: "PIRKANMAAN_ELY", sahkoposti: "kirjaamo.pirkanmaa@ely-keskus.fi" },
        { nimi: "POHJANMAAN_ELY", sahkoposti: "kirjaamo.pohjanmaa@ely-keskus.fi" },
        { nimi: "POHJOIS_KARJALAN_ELY", sahkoposti: "kirjaamo.pohjois-karjala@ely-keskus.fi" },
        { nimi: "POHJOIS_POHJANMAAN_ELY", sahkoposti: "kirjaamo.pohjois-pohjanmaa@ely-keskus.fi" },
        { nimi: "POHJOIS_SAVON_ELY", sahkoposti: "kirjaamo.pohjois-savo@ely-keskus.fi" },
        { nimi: "SATAKUNNAN_ELY", sahkoposti: "kirjaamo.satakunta@ely-keskus.fi" },
        { nimi: "UUDENMAAN_ELY", sahkoposti: "kirjaamo.uusimaa@ely-keskus.fi" },
        { nimi: "VARSINAIS_SUOMEN_ELY", sahkoposti: "kirjaamo.varsinais-suomi@ely-keskus.fi" },
        { nimi: "VAYLAVIRASTO", sahkoposti: "kirjaamo@vayla.fi" },
      ]);
    });

    describe("lisaaMuistutus", () => {
      it("should send email only to kirjaamo", async () => {
        const dbMockClient = mockClient(DynamoDBDocumentClient);
        const muistutusInput: MuistutusInput = {
          etunimi: "Mika",
          sukunimi: "Muistuttaja",
          katuosoite: "Muistojentie 1 a",
          postinumeroJaPostitoimipaikka: "00100 Helsinki",
          sahkoposti: undefined,
          puhelinnumero: "040123123",
          muistutus: "Hei. Haluaisin vain muistuttaa, että pihatieni yli täytyy rakentaa silta tai muu ratkaisu",
          liite: undefined,
        };

        await muistutusHandler.kasitteleMuistutus({ oid: fixture.PROJEKTI3_OID, muistutus: muistutusInput });

        sinon.assert.calledOnce(sendTurvapostiEmailStub);
        const calls = sendTurvapostiEmailStub.getCalls();
        expect(calls[0].args[0].to).to.equal("kirjaamo.uusimaa@ely-keskus.fi");
        expect(dbMockClient.commandCalls(UpdateCommand).length).to.equal(1);
        expect(dbMockClient.commandCalls(PutCommand).length).to.equal(1);
        const m = dbMockClient.commandCalls(PutCommand)[0].args[0].input.Item as DBMuistuttaja;
        expect(m.etunimi).to.equal("Mika");
        expect(m.sukunimi).to.equal("Muistuttaja");
        expect(m.sahkoposti).to.equal(undefined);
        expect(m.lahiosoite).to.equal("Muistojentie 1 a");
        expect(m.postinumero).to.equal("00100");
        expect(m.postitoimipaikka).to.equal("Helsinki");
        expect(m.puhelinnumero).to.equal("040123123");
        const updateCommand = dbMockClient.commandCalls(UpdateCommand)[0];
        assert(updateCommand.args[0].input.ExpressionAttributeValues);
        expect(updateCommand.args[0].input.ExpressionAttributeValues[":id"][0]).to.equal(m.id);
        assert(updateCommand.args[0].input.ExpressionAttributeNames);
        expect(updateCommand.args[0].input.ExpressionAttributeNames["#m"]).to.equal("muistuttajat");
      });

      it("should send emails to kirjaamo and muistuttaja successfully", async () => {
        mockClient(DynamoDBDocumentClient);
        const muistutusInput: MuistutusInput = {
          etunimi: "Mika",
          sukunimi: "Muistuttaja",
          katuosoite: undefined,
          postinumeroJaPostitoimipaikka: undefined,
          sahkoposti: "mika.muistuttaja@mikamuistutta.ja",
          puhelinnumero: undefined,
          muistutus: "Hei. Haluaisin vain muistuttaa, että pihatieni yli täytyy rakentaa silta tai muu ratkaisu",
          liite: undefined,
        };

        await muistutusHandler.kasitteleMuistutus({ oid: fixture.PROJEKTI3_OID, muistutus: muistutusInput });

        // Muistuttajalle sähköposti normaalia reittiä
        sinon.assert.callCount(sendEmailStub, 1);
        let calls = sendEmailStub.getCalls();
        expect(calls).to.have.length(1);
        expect(
          calls.map((call) => {
            return call.args[0].to;
          })
        ).to.have.members(["mika.muistuttaja@mikamuistutta.ja"]);

        // Kirjaamoon sähköposti turvapostin kautta
        sinon.assert.callCount(sendTurvapostiEmailStub, 1);
        calls = sendTurvapostiEmailStub.getCalls();
        expect(calls).to.have.length(1);
        expect(
          calls.map((call) => {
            return call.args[0].to;
          })
        ).to.have.members(["kirjaamo.uusimaa@ely-keskus.fi"]);
      });
    });
    describe("haeMuistuttajat", () => {
      beforeEach(() => {
        loadProjektiByOidStub.restore();
        identifyMockUser({ etunimi: "", sukunimi: "", uid: "testuid", __typename: "NykyinenKayttaja" });
      });
      it("should get muistuttajat", async () => {
        const dbMock = mockClient(DynamoDBDocumentClient);
        dbMock
          .on(GetCommand, { TableName: config.projektiTableName })
          .resolves({ Item: { id: "1", muistuttajat: ["1", "2"], kayttoOikeudet: [{ kayttajatunnus: "testuid" }] } });
        dbMock.on(BatchGetCommand).resolves({
          Responses: {
            [config.muistuttajaTableName]: [
              {
                id: "1",
                etunimi: "Matti",
                sukunimi: "Teppo",
                lahiosoite: "Osoite 1",
                postinumero: "00100",
                postitoimipaikka: "Helsinki",
              },
              {
                id: "2",
                etunimi: "Teppo",
                sukunimi: "Testaaja",
                lahiosoite: "Osoite 2",
                postinumero: "01000",
                postitoimipaikka: "Vantaa",
              },
            ],
          },
        });
        const muistuttajat = await muistutusHandler.haeMuistuttajat({ oid: "1.2.3", muutMuistuttajat: false, sivu: 1, sivuKoko: 1 });
        expect(dbMock.commandCalls(BatchGetCommand).length).to.be.equal(1);
        let batchCommand = dbMock.commandCalls(BatchGetCommand)[0];
        assert(batchCommand.args[0].input.RequestItems);
        let keys = batchCommand.args[0].input.RequestItems[config.muistuttajaTableName].Keys;
        assert(keys);
        expect(keys.length).to.be.equal(1);
        expect(keys[0].id).to.be.equal("1");
        expect(muistuttajat.hakutulosMaara).to.equal(2);
        expect(muistuttajat.muistuttajat[0]?.etunimi).to.equal("Matti");
        expect(muistuttajat.muistuttajat[0]?.sukunimi).to.equal("Teppo");
        expect(muistuttajat.muistuttajat[0]?.jakeluosoite).to.equal("Osoite 1");
        expect(muistuttajat.muistuttajat[0]?.postinumero).to.equal("00100");
        expect(muistuttajat.muistuttajat[0]?.paikkakunta).to.equal("Helsinki");
        await muistutusHandler.haeMuistuttajat({ oid: "1.2.3", muutMuistuttajat: false, sivu: 2, sivuKoko: 1 });
        batchCommand = dbMock.commandCalls(BatchGetCommand)[1];
        assert(batchCommand.args[0].input.RequestItems);
        keys = batchCommand.args[0].input.RequestItems[config.muistuttajaTableName].Keys;
        assert(keys);
        expect(keys.length).to.be.equal(1);
        expect(keys[0].id).to.be.equal("2");
        expect(muistuttajat.hakutulosMaara).to.equal(2);
      });
      it("should get muut muistuttajat", async () => {
        const dbMock = mockClient(DynamoDBDocumentClient);
        dbMock
          .on(GetCommand, { TableName: config.projektiTableName })
          .resolves({ Item: { id: "1", muutMuistuttajat: ["11"], kayttoOikeudet: [{ kayttajatunnus: "testuid" }] } });
        dbMock.on(BatchGetCommand).resolves({
          Responses: {
            [config.muistuttajaTableName]: [
              {
                id: "11",
                etunimi: "Matti",
                sukunimi: "Teppo",
                lahiosoite: "Osoite 1",
                postinumero: "00100",
                postitoimipaikka: "Helsinki",
              },
            ],
          },
        });
        const muistuttajat = await muistutusHandler.haeMuistuttajat({ oid: "1.2.3", muutMuistuttajat: true, sivu: 1, sivuKoko: 1 });
        expect(dbMock.commandCalls(BatchGetCommand).length).to.be.equal(1);
        const batchCommand = dbMock.commandCalls(BatchGetCommand)[0];
        assert(batchCommand.args[0].input.RequestItems);
        const keys = batchCommand.args[0].input.RequestItems[config.muistuttajaTableName].Keys;
        assert(keys);
        expect(keys.length).to.be.equal(1);
        expect(keys[0].id).to.be.equal("11");
        expect(muistuttajat.hakutulosMaara).to.equal(1);
      });
    });
    describe("tallennaMuistuttajat", () => {
      beforeEach(() => {
        loadProjektiByOidStub.restore();
        identifyMockUser({ etunimi: "", sukunimi: "", uid: "testuid", __typename: "NykyinenKayttaja" });
      });
      it("should save muistuttajat", async () => {
        const dbMock = mockClient(DynamoDBDocumentClient);
        dbMock
          .on(GetCommand, { TableName: config.projektiTableName })
          .resolves({ Item: { id: "1", muutMuistuttajat: ["11"], kayttoOikeudet: [{ kayttajatunnus: "testuid" }] } });
        dbMock
          .on(GetCommand, { TableName: config.muistuttajaTableName, Key: { id: "1" } })
          .resolves({ Item: { id: "1", nimi: "Matti Teppo" } });
        const muistuttajat = await muistutusHandler.tallennaMuistuttajat({
          oid: "1.2.3",
          muistuttajat: [
            { id: "1", nimi: "Matti Teppo", tiedotusosoite: "matti@teppo.fi", tiedotustapa: "email" },
            { nimi: "Teppo Testaaja", tiedotusosoite: "test@test.fi" },
          ],
        });
        expect(dbMock.commandCalls(PutCommand).length).to.be.equal(2);
        const putCommand = dbMock.commandCalls(PutCommand)[0];
        const o1 = putCommand.args[0].input.Item as DBMuistuttaja;
        expect(o1.nimi).to.be.equal("Matti Teppo");
        expect(o1.tiedotusosoite).to.be.equal("matti@teppo.fi");
        expect(o1.tiedotustapa).to.be.equal("email");
        const putCommand2 = dbMock.commandCalls(PutCommand)[1];
        const o2 = putCommand2.args[0].input.Item as DBMuistuttaja;
        expect(o2.nimi).to.be.equal("Teppo Testaaja");
        expect(o2.tiedotusosoite).to.be.equal("test@test.fi");
        expect(o2.tiedotustapa).to.be.equal(undefined);
        expect(dbMock.commandCalls(UpdateCommand).length).to.be.equal(1);
        const updateCommand = dbMock.commandCalls(UpdateCommand)[0];
        assert(updateCommand.args[0].input.ExpressionAttributeValues);
        expect(updateCommand.args[0].input.ExpressionAttributeValues[":id"].length).to.be.equal(1);
        assert(updateCommand.args[0].input.ExpressionAttributeNames);
        expect(updateCommand.args[0].input.ExpressionAttributeNames["#m"]).to.equal("muutMuistuttajat");
        expect(muistuttajat.length).to.equal(2);
        expect(muistuttajat[0].id).to.equal("1");
        expect(muistuttajat[0].nimi).to.equal("Matti Teppo");
        expect(muistuttajat[0].tiedotusosoite).to.equal("matti@teppo.fi");
        expect(muistuttajat[0].tiedotustapa).to.equal("email");
        expect(muistuttajat[1].id !== undefined).to.equal(true);
        expect(muistuttajat[1].nimi).to.equal("Teppo Testaaja");
        expect(muistuttajat[1].tiedotusosoite).to.equal("test@test.fi");
        expect(muistuttajat[1].tiedotustapa).to.equal(undefined);
      });
    });
    describe("poistaMuistuttaja", () => {
      beforeEach(() => {
        loadProjektiByOidStub.restore();
        identifyMockUser({ etunimi: "", sukunimi: "", uid: "testuid", __typename: "NykyinenKayttaja" });
      });
      it("should delete muistuttaja", async () => {
        const dbMock = mockClient(DynamoDBDocumentClient);
        dbMock.on(GetCommand).resolves({ Item: { id: "1", muistuttajat: ["1"], muutMuistuttajat: ["2", "3"], kayttoOikeudet: [{ kayttajatunnus: "testuid"}] } });
        await muistutusHandler.poistaMuistuttaja({ oid: "1", muistuttaja: "2" });
        expect(dbMock.commandCalls(UpdateCommand).length).to.be.equal(1);
        const updateCommand = dbMock.commandCalls(UpdateCommand)[0];
        assert(updateCommand.args[0].input.ExpressionAttributeValues);
        expect(updateCommand.args[0].input.ExpressionAttributeValues[":muutMuistuttajat"].length).to.be.equal(1);
        expect(updateCommand.args[0].input.ExpressionAttributeValues[":muutMuistuttajat"][0]).to.be.equal("3");
      });
    });
  });
});
