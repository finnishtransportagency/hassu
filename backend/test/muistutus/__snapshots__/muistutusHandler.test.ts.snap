// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`muistutusHandler handleEvent lisaaMuistutus should send email only to kirjaamo 1`] = `
Object {
  "subject": "Muistutus (VLS) Mika Muistuttaja ELY/3/2022",
  "text": "Muistutus (VLS) Mika Muistuttaja ELY/3/2022

Vahvistus muistutuksen jättämisestä Valtion liikenneväylien suunnittelu -järjestelmän kautta
Aihe:
Muistutus on vastaanotettu
09.04.2024 klo 03:00
Suunnitelman nimi
Testiprojekti 3
Suunnitelman asiatunnus
ELY/3/2022

Muistutuksen lähettäjän tiedot
Nimi
Mika Muistuttaja

Osoite
Muistojentie 1 a
00100 Helsinki

Sähköposti
test@test.fi
Puhelinnumero
0501234567
Muistutus
Hei. Haluaisin vain muistuttaa, että pihatieni yli täytyy rakentaa silta tai muu ratkaisu

Suunnitelman tietoihin pääset tästä linkistä: https://localhost:3000/suunnitelma/3/nahtavillaolo",
  "to": "kirjaamo.uusimaa@ely-keskus.fi",
}
`;
